libr/bin/format/elf/elf.c

```diff
@@ -1277,228 +1299,185 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
+	for (i = 0; i < bin->reloc_num; i++) {
+		RBinElfReloc *tmp = R_NEW (RBinElfReloc);
+		memcpy (tmp, relocs + i, sizeof (RBinElfReloc));
+		if (!ht_up_insert (rel_cache, tmp->sym, tmp)) {
+			free (tmp);
+		}
+	}
+	free (relocs);
+	return rel_cache;
 }
 
-static struct ht_rel_t *read_ht_rel(ELFOBJ *bin, ut8 *rl, int k) {
-	struct ht_rel_t *rel = R_NEW0 (struct ht_rel_t);
-	if (!rel) {
-		return NULL;
+static ut64 get_got_entry(ELFOBJ *bin, RBinElfReloc *rel) {
+	if (!rel->rva) {
+		return UT64_MAX;
 	}
 
-	rel->is_rela = bin->is_rela == DT_RELA;
-	rel->k = k;
-	if (rel->is_rela) {
-		read_rela (bin, &rel->r.rela, rl);
-	} else {
-		read_rel (bin, &rel->r.rel, rl);
+	ut64 p_sym_got_addr = Elf_(r_bin_elf_v2p_new) (bin, rel->rva);
+
+	ut8 buf[WORDSIZE];
+	int res = r_buf_read_at (bin->b, p_sym_got_addr, buf, WORDSIZE);
+
+	if (res != WORDSIZE) {
+		return UT64_MAX;
+	}
+
+	size_t i = 0;
+	ut64 addr = READWORD (buf, i);
+
+	if (!addr) {
+		return UT64_MAX;
 	}
-	return rel;
+
+	return addr;
 }
 
-static void rel_cache_free(HtUPKv *kv) {
-	free (kv->value);
+static bool is_thumb_symbol(ut64 plt_addr) {
+	return plt_addr & 1;
 }
 
-static HtUP *rel_cache_new(ELFOBJ *bin) {
-	ut8 rl[MAX_REL_RELA_SZ] = { 0 };
-	RBinElfSection *rel_sec = NULL;
-	int j, k, tsize, nrel;
-	const char *rel_sect[] = { ".rel.plt", ".rela.plt", ".rel.dyn", ".rela.dyn", NULL };
-	const char *rela_sect[] = { ".rela.plt", ".rel.plt", ".rela.dyn", ".rel.dyn", NULL };
-	HtUP *rel_cache;
+static ut64 get_import_addr_arm(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 got_addr = bin->dyn_info.dt_pltgot;
+	if (!got_addr) {
+		return UT64_MAX;
+	}
 
-	if ((!bin->shdr || !bin->strtab) && !bin->phdr) {
-		return NULL;
+	ut64 plt_addr = get_got_entry (bin, rel);
+	if (plt_addr == UT64_MAX) {
+		return UT64_MAX;
 	}
 
-	if (bin->is_rela == DT_REL) {
-		rel_sec = get_rel_sec (bin, rel_sect);
-		tsize = sizeof (Elf_(Rel));
-	} else if (bin->is_rela == DT_RELA) {
-		rel_sec = get_rel_sec (bin, rela_sect);
-		tsize = sizeof (Elf_(Rela));
+	ut64 pos = (rel->rva - got_addr - 0x3 * WORDSIZE) / WORDSIZE;
+
+	switch (rel->type) {
+	case R_ARM_JUMP_SLOT: {
+		plt_addr += pos * 12 + 20;
+		if (is_thumb_symbol (plt_addr)) {
+			plt_addr--;
+		}
+		return plt_addr;
 	}
-	if (!rel_sec || rel_sec->size < 1) {
-		return NULL;
+	case R_AARCH64_JUMP_SLOT:
+		return plt_addr + pos * 16 + 32;
+	default:
+		bprintf ("Unsupported relocation type for imports %d\n", rel->type);
+		return UT64_MAX;
 	}
+	return UT64_MAX;
+}
 
-	nrel = (ut32) ((int)rel_sec->size / (int)tsize);
-	if (nrel < 1) {
-		return NULL;
+static ut64 get_import_addr_mips(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 addr_jmprel = bin->dyn_info.dt_jmprel;
+	ut64 addr_got = bin->dyn_info.dt_mips_pltgot;
+
+	if (!addr_jmprel || !addr_got) {
+		return UT64_MAX;
 	}
 
-	const int htsize = R_MIN (nrel, 1024);
-	rel_cache = ht_up_new_size (htsize, NULL, rel_cache_free, NULL);
+	ut64 pos = (rel->rva - addr_got - 0x2 * WORDSIZE) / WORDSIZE;
```

Comment:
I still think it would be much easier to just store the position from the very beginning, instead of having to computing it every time with "hard" computations.